main.c

```diff
@@ -360,12 +375,47 @@ STATIC bool run_code_py(safe_mode_t safe_mode) {
 
         // Refresh the ePaper display if we have one. That way it'll show an error message.
         #if CIRCUITPY_DISPLAYIO
+        // Don't refresh the display if we're about to deep sleep.
+        #if CIRCUITPY_ALARM
+        refreshed_epaper_display = refreshed_epaper_display || result.return_code & PYEXEC_DEEP_SLEEP;
+        #endif
         if (!refreshed_epaper_display) {
             refreshed_epaper_display = maybe_refresh_epaperdisplay();
         }
         #endif
 
-        tick_rgb_status_animation(&animation);
+        // Sleep until our next interrupt.
+        #if CIRCUITPY_ALARM
+        if (result.return_code & PYEXEC_DEEP_SLEEP) {
+            // Make sure we have been awake long enough for USB to connect (enumeration delay).
+            int64_t connecting_delay_ticks = CIRCUITPY_USB_CONNECTED_SLEEP_DELAY * 1024 - port_get_raw_ticks(NULL);
+            if (connecting_delay_ticks > 0) {
+                // Set when we've waited long enough so that we wake up from the
+                // sleep_until_interrupt below and loop around to the real deep
+                // sleep in the else clause.
+                port_interrupt_after_ticks(connecting_delay_ticks);
+            // Deep sleep if we're not connected to a host.
+            } else if (!asleep) {
+                asleep = true;
+                new_status_color(BLACK);
+                board_deinit();
+                if (!supervisor_workflow_active()) {
+                    // Enter true deep sleep. When we wake up we'll be back at the
+                    // top of main(), not in this loop.
+                    alarm_enter_deep_sleep();
+                    // Does not return.
+                } else {
+                    serial_write_compressed(translate("Pretending to deep sleep until alarm, any key or file write.\n"));
+                }
+            }
+        }
+        #endif
+
+        if (!asleep) {
+            tick_rgb_status_animation(&animation);
+        } else {
+            port_idle_until_interrupt();
```

Comment:
Maybe add a comment here about when this is going to wait. I know it will wait if we haven't reached the enumeration delay yet, but does it also wait on fake sleep?